{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ddb5fbd-7dec-4cfc-bc67-6d641d119d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config InlineBackend.figure_format = 'retina'\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "plt.rcParams['pdf.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb297fe0-2588-420c-b6e8-9627ec28e016",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genome and annotation\n",
    "\n",
    "from src.DMS_Profile import Genome, Annotation\n",
    "\n",
    "genome = Genome('/Users/leo/Documents/repos/lsStruct/data/genome/scer_chr_kan.FASTA')\n",
    "annotation = Annotation('/Users/leo/Documents/repos/lsStruct/data/genome/scer_kan.gff')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f763dfe0-269f-4636-9af1-90888461fe76",
   "metadata": {},
   "source": [
    "## Figure 1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1758bac9-77de-44aa-b1b0-cdedead3bff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.DMS_Profile import PRO_Profile, pro_combine_profiles\n",
    "from src.pro_utils import moving_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90dc208a-7dd9-48fd-9e39-89b968426913",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "\n",
    "spt4_pro = pro_combine_profiles([PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_spt4_1_mRNA_pro.pkl', 'spt4_R1', genome, norm=False), \n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_spt4_2_mRNA_pro.pkl', 'spt4_R2', genome, norm=False), \n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_spt4_3_mRNA_pro.pkl', 'spt4_R3', genome, norm=False)], 'spt4', genome, norm=True)\n",
    "\n",
    "wild_pro = pro_combine_profiles([PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_1_mRNA_pro.pkl', 'wild_R1', genome, norm=False), \n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_2_mRNA_pro.pkl', 'wild_R2', genome, norm=False),\n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_3_mRNA_pro.pkl', 'wild_R3', genome, norm=False),\n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_4_mRNA_pro.pkl', 'wild_R4', genome, norm=False), \n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_5_mRNA_pro.pkl', 'wild_R5', genome, norm=False),\n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_6_mRNA_pro.pkl', 'wild_R6', genome, norm=False),\n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_7_mRNA_pro.pkl', 'wild_R7', genome, norm=False), \n",
    "                                 PRO_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_pro/pro_wild_8_mRNA_pro.pkl', 'wild_R8', genome, norm=False)], 'wild', genome, norm=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0259f80e-46b6-401f-ba80-57332fa1cac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot individual profile\n",
    "\n",
    "fig, axs = plt.subplots(2, 1, figsize=[6, 2], sharex=True, sharey=True)\n",
    "plt.subplots_adjust(hspace=0.3)\n",
    "\n",
    "avg_win = 5\n",
    "\n",
    "# RPS17A\n",
    "chrom = 12\n",
    "coords = (225889, 226697)\n",
    "\n",
    "wild_pro.plot_profile(chrom, coords[0], coords[1], genome, axs[0], avg_win=avg_win)\n",
    "spt4_pro.plot_profile(chrom, coords[0], coords[1], genome, axs[1], avg_win=avg_win)\n",
    "\n",
    "plt.setp(axs, ylim=[0, 4])\n",
    "plt.xlabel('Genomic coordinate')\n",
    "plt.ylabel('Pol II per million')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f847bec2-aa9f-4296-b923-7b96d47098c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot meta profile\n",
    "\n",
    "avg_win = 8 # should be even\n",
    "n_bins  = 100\n",
    "surr    = 0.1\n",
    "min_ppm = 1\n",
    "\n",
    "plt.figure(figsize=(3, 2))\n",
    "\n",
    "meta_wild, norm_x, bin_size = wild_pro.get_meta_profile_binned(annotation, n_bins=n_bins, surr=surr)\n",
    "meta_spt4, norm_x, bin_size = spt4_pro.get_meta_profile_binned(annotation, n_bins=n_bins, surr=surr)\n",
    "\n",
    "mask_wild = np.array([np.all(i == i) for i in meta_wild])\n",
    "mask_spt4 = np.array([np.all(i == i) for i in meta_spt4])\n",
    "\n",
    "meta_wild_mean = np.mean(meta_wild[mask_wild & mask_spt4], axis=0)\n",
    "meta_spt4_mean = np.mean(meta_spt4[mask_wild & mask_spt4], axis=0)\n",
    "\n",
    "plt.plot(norm_x, meta_wild_mean, 'or', ms=5, mew=0)\n",
    "plt.plot(norm_x, meta_spt4_mean, 'og', ms=5, mew=0)\n",
    "\n",
    "plt.plot(norm_x[int(avg_win/2):-int(avg_win/2)+1], moving_average(meta_wild_mean, avg_win), '-r', lw=2, label=f\"wildtype\")\n",
    "plt.plot(norm_x[int(avg_win/2):-int(avg_win/2)+1], moving_average(meta_spt4_mean, avg_win), '-g', lw=2, label=f\"âˆ†STP4\")\n",
    "\n",
    "plt.plot([surr, surr], [np.min(meta_spt4_mean), np.max(meta_spt4_mean)], '--k')\n",
    "plt.plot([1-surr, 1-surr], [np.min(meta_spt4_mean), np.max(meta_spt4_mean)], '--k')\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel('normalized Pol II position in gene')\n",
    "plt.ylabel('mean Pol II per million')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3041591d-0073-425f-8262-5eb0ecbd3e10",
   "metadata": {},
   "source": [
    "## Figure 1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b100c80-cb22-44ff-ae4e-4e8c29376211",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.DMS_Profile import DMS_Profile, combine_profiles, gini_for_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf5a65f6-7a80-443a-80e5-e977f6642dc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load DMS data\n",
    "\n",
    "dms_pro = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_2_mRNA_agg.pkl', 'wild_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_3_mRNA_agg.pkl', 'wild_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_3_mRNA_agg.pkl', 'wild_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_4_mRNA_agg.pkl', 'wild_R4', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_5_mRNA_agg.pkl', 'wild_R5', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_6_mRNA_agg.pkl', 'wild_R6', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_7_mRNA_agg.pkl', 'wild_R7', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_8_mRNA_agg.pkl', 'wild_R8', genome)], 'all_pro', genome, min_cov=900)\n",
    "\n",
    "dms_mat = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_2_mRNA_agg.pkl', 'wild_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_3_mRNA_agg.pkl', 'wild_R3', genome)], 'all_mat', genome, min_cov=900)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd2b534-d431-4935-b158-ef79632b21a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot individual profile for DBP2\n",
    "\n",
    "chrom = 13\n",
    "coords = (414760, 415000)\n",
    "strand = '+'\n",
    "\n",
    "fig, axs = plt.subplots(2, 1, figsize=[15, 2], sharex=True)\n",
    "plt.subplots_adjust(hspace=0.1)\n",
    "\n",
    "ax0 = axs[0].twinx()\n",
    "dms_pro.plot_profile(chrom, strand, coords[0], coords[1], genome, axs[0], cmap_loc='../cmap.txt')\n",
    "ax0.bar(np.arange(coords[0], coords[1]), dms_pro.cov_p[chrom][coords[0]:coords[1]], 1, color='lightgray')\n",
    "axs[0].set_zorder(10)\n",
    "axs[0].set_frame_on(False)\n",
    "\n",
    "ax1 = axs[1].twinx()\n",
    "dms_mat.plot_profile(chrom, strand, coords[0], coords[1], genome, axs[1], cmap_loc='../cmap.txt')\n",
    "ax1.bar(np.arange(coords[0], coords[1]), dms_mat.cov_p[chrom][coords[0]:coords[1]], 1, color='lightgray')\n",
    "axs[1].set_zorder(10)\n",
    "axs[1].set_frame_on(False)\n",
    "axs[1].set_ylabel('DMS Reactivity')\n",
    "axs[1].set_xlabel('Genomic coordinate')\n",
    "\n",
    "plt.setp([axs[0], axs[1]], ylim=(0, 2.5))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bebb8e40-efc3-4c2b-b618-5db39f595585",
   "metadata": {},
   "source": [
    "## Figure 1d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a76b98d-d176-42b4-90a4-80aeef10469d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.DMS_Profile import Targeted_DMS_Profile, targeted_combine_profiles\n",
    "from src.dms_utils import sliding_corrcoeff, gini\n",
    "from scipy.stats import linregress\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from src.pydscatter import dscatter_plot\n",
    "import matplotlib as mpl\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mplc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1daa5c6b-000c-4a2c-91bc-c71db20ab4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_corr_vec(rea1, rea2, mask, w):\n",
    "    # wrapper for sliding_corrcoeff() to handle NaNs\n",
    "    mask_join = (rea1 == rea1) & (rea2 == rea2) & mask\n",
    "    r = sliding_corrcoeff(rea1[mask_join], rea2[mask_join], w)\n",
    "    _, _, r_value, _, _ = linregress(rea1[mask_join], rea2[mask_join])\n",
    "    x = np.arange(1, len(mask)+1)[mask_join]\n",
    "    return(x, r, r_value)\n",
    "\n",
    "def gini_windows(rea, mask, w, min_n=20):\n",
    "    ginis = []\n",
    "    x = []\n",
    "    for window in range(len(rea)-w):\n",
    "        if len((rea[window:window+w])[mask[window:window+w]]) >= min_n:\n",
    "            gini_w = gini((rea[window:window+w])[mask[window:window+w]])\n",
    "            ginis.append(gini_w)\n",
    "        else:\n",
    "            ginis.append(np.nan)\n",
    "        x.append((window+window+w)/2)\n",
    "    return(x, ginis)\n",
    "\n",
    "def get_partner(dbr):\n",
    "    bpdict = {}\n",
    "    opened = []\n",
    "    for i, d in enumerate(dbr):\n",
    "        if d == '(':\n",
    "            opened.append(i)\n",
    "        elif d == ')':\n",
    "            partner = opened.pop(-1)\n",
    "            if i not in bpdict:\n",
    "                bpdict[i] = partner\n",
    "                bpdict[partner] = i\n",
    "    \n",
    "    if opened:\n",
    "        print('parentheses unmatched!')\n",
    "        return(None)\n",
    "    else:\n",
    "        return(bpdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6f8254-4973-42b4-a2a4-620f10095726",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load nascent and mature average rRNA data\n",
    "genome_rrna = Genome('/Users/leo/Documents/repos/lsStruct/data/genome/scer_chr_kan.FASTA', coords=(11, 451575, 458433), reverse=True)\n",
    "\n",
    "wild_rib_1 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/rib_wild_1_rRNA_agg.pkl', 'wild_R1', genome_rrna, reverse=True)\n",
    "wild_rib_2 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/rib_wild_2_rRNA_agg.pkl', 'wild_R2', genome_rrna, reverse=True) \n",
    "wild_rib_3 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/rib_wild_3_rRNA_agg.pkl', 'wild_R3', genome_rrna, reverse=True)\n",
    "wild_rib = targeted_combine_profiles([wild_rib_1, wild_rib_2, wild_rib_3], 'wild', genome_rrna, per_exclude=5)\n",
    "\n",
    "wild_nas_1 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_1_rRNA_agg.pkl', 'wild_R1', genome_rrna, reverse=True) \n",
    "wild_nas_2 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_2_rRNA_agg.pkl', 'wild_R2', genome_rrna, reverse=True)\n",
    "wild_nas_3 = Targeted_DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_3_rRNA_agg.pkl', 'wild_R3', genome_rrna, reverse=True)\n",
    "wild_nas = targeted_combine_profiles([wild_nas_1, wild_nas_2, wild_nas_3], 'wild', genome_rrna, per_exclude=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f43a938-9d66-49d4-986f-2508d977d048",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ribosomal RNA secondary structure\n",
    "dbr_18 = '...((((.........))))((((.(((((((.(((((((((.....(((.(((..((...(((..(.((..........)))..))))).....((((.......(((((((..((..(((((((............(((((....(((((((.....)))))))....)))))......(((((((((....)))))))))(((.(((((((.......(((((.(((....)))...))))).....))))))).)..))...((((.((((.....))))))))..))))))).))))))))).(((..(.(((....((((((((.......))))))))))).....))))...((((((((....))))...))))))))((((((..........)))))).((((....))))...)))))))......(.(((...(((((...))))).)))).)).))))))....((((..(((((((....)))))))..).))).....((((((((.......))))))))........((.((......(.((((((..(((....)))....))))))))).)).))))))))))).....(...(((.......((((...(((.((....((((((((((((.((((.(((.....)))...)))).....))))))))))))....((((((....(((((((((.....)))))))))....))))))..(((((((((.......(((..(.(...).)..(((.....)))...)))......)))))..)))).....(.((....(.((.(((.............))).))..)..)).)..))...((((((((((.((((((((((((((((((((...(((......)))......))))))))))))....(..((....)))))))))))))))).))))..))))...)))).(..((((((...(((.(((((.........))))).)))))))))..).......((((((.(((..(((((((...((..........)))))))))..)))...((....))...)))....))).))))(((((.((.((((....)))))))))))........(((((.(((((((..((..(((((((((((((((((.(.)((((........))))........(((((((....(((((....(((((((((..........)))))))))..))))).(.((.((((..((((((((((..(((((((((....)))..((((......))))..)))))).....((((((((.((((..(((((.((((((....))))))...)))))..))))))).((.(((((((...)))))))))....)))))...))))).)))...).))))))))....)))))))...)).)))))))))((..(((((((.(...(((.....(((.((((....)))).)))....)))....).)))))))....).((((((((((((........))))))))))))..).))))))(...(((((((((.......)))))))))..)..))...)))))))))).))....((.((...(((((((((((.((((((((((((..(((((((((((((((((((((((((((....)))))))))))))))))))))))))))..)))))))))))))))))))))))....))..))....((((((((((....))))))))))........'\n",
    "dbr_58 = '........................................((((((((......)))).....((((((....................)))).))...))))...(((...)))(((((((((....))))))))).....................'\n",
    "dbr_25 = '..........................((((((.....((....))......))))).).........((....))..(((((......((.....)).....))))).((((..((((((((((((((((((...))))))))))))))))))....(((((((.((((((((((((((((((((...(.(((((....))))))..(...).(((.......)))....))))))))..))))))))....)))))))))))........(((((((........))))))).....(((((........)))))..)))).................................((....))...............((((....))))..................................((((((((((((.(..(((((...((((..((((((((.....))))))))......))))...))))).(((((((((((((((((((((((......((((((((..((((((((((....))))))..)))).))))))))....))))).))))))))))).......(((....))).....)))))))..).))))))))))))..((((........))))((((((....(((((((((((((..((((..((((.....))))..))))...((....)).......(((((..(((((....)))))...))))).((((((((...((....))...))).))).))....)))))))))))))..((((...((((((...((((((((((((.((((...(((((((....)))))))....))))...((((.((.((........))))))))..).))))))))))).(((....))).......))))))......))))..((((((....(((((....))))).(((((((((.......((((((((..(.(((....((((((((((..((((((((.....))))))))))))))))))..)).).)..)))))))).(((((((((.........)))))))))...........)))))))).)...((((((((((.....))))))))))....(((......)))...(((((((...(((......(.((((((.................(((.((((((((((.......(((((.(((..(((.........)))..)))(...)(...((......)).).)))))...)))))))))))))...................))))))))))...)))))))(((((((((.........)))))))))....))))))(.((((((((.((......((((.((((....))))))))...))))))))))..).....))))))((((.((((.....((((((.......))))))((((.....(((((.....(((((((...((.....)).))))))).....(((.((...(((((.(((....))))))......(((((........))))).....))....))...(((((........)))))..((((.(((((((.....(((....))).......((((((...((((.(((((((.(((((((((.........))))))...(((((((..(((((((((((((.....)))))))))))))..)).)))))......(((((((....)))))))..)))...))))))).)))).........))))))............))))))))))).))).........((......))))))).((((.....))))..))))..((((....((((((((((((..............(((((...((....))....)))))((((((((((((((.((((((((((((((((((....(((((((((((((((((((.(((((((((((((....))))))))))))).))))))))))))).))))))....))))))))))(((((.....)))))....))))))))......)))))..)))))))))(...).......(((((((((..(((.........)))..(((((((((..((((.....))))..)..))))))))..((((.(((..((((((.(..(((....(((((....)))))....))))))))..)))))((((((.......))))))(...).((((.(...(.)).(((((((.)..))))))..)..)))..))))...)).)))))))....)))))))))))).))))))))))))....(((((........)))))..((((((((.(((.((.....(((((.(((((((.(((((..(((((((..((((((((((((((((((.........(((((..((....))............))))))))))))).))))))))))...(((((.....((((..(((((......)))))...............((((((....))))))..))))..))))).....)))))))...)))))..((((.....))).)((((((...........))))))..((..((((((((((.((((((.......))))))...))((......))....))))))))..(((((((((.......))))))))).(((....))).))..........((((((..((((.......)))))))))).........)))))..))))))).......(((..(((((((...((.......))...)))))))...)))..........(((((((((((..((((((((((....))))))))..)).(((((.....))))).....)))))..)..)))).).....(((((((....))).))))....))..)))))))))))..((((((((..((((.((((((((...............))))))))(((((...(((....(((((((((((....))))))..)))))...)))..)))))..(((....(((((..........)))))....))).))))....)))))))).......(((((((((((((................((((((.....))))))...............((((((((..(((((((((((......))))))))))))))))))).............)))))))))))))..((((((((((.......)))))....(((((((((.((..(..((((.((((((((......)))))))))))))...))....))).))))))..)))))..'\n",
    "dbr_full = 701*'?' + dbr_18 + 361*'?' + dbr_58 + 232*'?' + dbr_25 + 210*'?'\n",
    "mask_dat = np.array([True if i in [')', '(', '.'] else False for i in dbr_full])\n",
    "\n",
    "## calculate sliding correlation coefficients\n",
    "mask_acu = genome_rrna.mask_A()|genome_rrna.mask_C()|genome_rrna.mask_U() # joint mask for mature and nascent RNA\n",
    "\n",
    "w = 50\n",
    "x_nasvmat, r_nasvmat, r2_nasvmat = get_corr_vec(wild_nas.rat, wild_rib.rat, mask_acu, w)\n",
    "x_nas,     r_nas    , r2_nas     = get_corr_vec(wild_nas_1.rat, wild_nas_2.rat, mask_acu, w)\n",
    "x_mat,     r_mat    , r2_mat = get_corr_vec(wild_rib_1.rat, wild_rib_2.rat, mask_acu, w)\n",
    "\n",
    "## calculate sliding gini index\n",
    "w = 80\n",
    "x_gini_nas, gini_nas = gini_windows(wild_nas.rat, mask_acu&wild_nas.mask, w, min_n=50)\n",
    "x_gini_mat, gini_mat = gini_windows(wild_rib.rat, mask_acu&wild_rib.mask, w, min_n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b85a55-9556-4101-953a-20396d7905b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(layout=\"constrained\", figsize=(13,4))\n",
    "\n",
    "gs = GridSpec(4, 2, figure=fig, width_ratios=(0.88, 0.12))\n",
    "ax1 = fig.add_subplot(gs[0, 0])\n",
    "ax2 = fig.add_subplot(gs[1, 0], sharex=ax1)\n",
    "ax3 = fig.add_subplot(gs[2, 0], sharex=ax1)\n",
    "ax4 = fig.add_subplot(gs[0, 1])\n",
    "ax5 = fig.add_subplot(gs[1, 1])\n",
    "ax6 = fig.add_subplot(gs[2, 1])\n",
    "ax7 = fig.add_subplot(gs[3, 0], sharex=ax1)\n",
    "ax8 = fig.add_subplot(gs[3, 1])\n",
    "\n",
    "cmap = np.array(pd.read_csv('../cmap.txt', header=None))/255\n",
    "cmap = mplc.ListedColormap(cmap)\n",
    "\n",
    "ax1.fill_between(np.arange(0, len(mask_dat)), mask_dat, color='lightgray')\n",
    "ax2.fill_between(np.arange(0, len(mask_dat)), mask_dat, color='lightgray')\n",
    "ax3.fill_between(np.arange(0, len(mask_dat)), mask_dat, color='lightgray')\n",
    "ax1.plot(x_nasvmat, r_nasvmat, '.')\n",
    "ax2.plot(x_nas, r_nas, '.')\n",
    "ax2.set_ylabel(\"Pearson's correlation\")\n",
    "ax3.plot(x_mat, r_mat, '.')\n",
    "\n",
    "dscatter_plot(np.log10(wild_rib.rat[mask_acu]), np.log10(wild_nas.rat[mask_acu]), ax=ax4, order=True, alpha=0.15, markersize=5, edgecolors='none', rasterized=True, cmap=cmap)\n",
    "ax4.set_aspect('equal', adjustable=\"datalim\")\n",
    "ax4.set_xticks(np.arange(-6, 1, 1))\n",
    "ax4.set_yticks(np.arange(-4, 1, 1))\n",
    "ax4.text(0.005, 0.82, f\"R$={r2_nasvmat:.3f}\", transform=ax4.transAxes)\n",
    "dscatter_plot(np.log10(wild_nas_1.rat[mask_acu]), np.log10(wild_nas_2.rat[mask_acu]), ax=ax5, order=True, alpha=0.15, markersize=5, edgecolors='none', rasterized=True, cmap=cmap)\n",
    "ax5.set_aspect('equal', adjustable=\"datalim\")\n",
    "ax5.set_xticks(np.arange(-6, 1, 1))\n",
    "ax5.set_yticks(np.arange(-4, 1, 1))\n",
    "ax5.text(0.005, 0.82, f\"R$={r2_nas:.3f}\", transform=ax5.transAxes)\n",
    "dscatter_plot(np.log10(wild_rib_1.rat[mask_acu]), np.log10(wild_rib_2.rat[mask_acu]), ax=ax6, order=True, alpha=0.15, markersize=5, edgecolors='none', rasterized=True, cmap=cmap)\n",
    "ax6.set_aspect('equal', adjustable=\"datalim\")\n",
    "ax6.set_xticks(np.arange(-10, 2, 2))\n",
    "ax6.set_yticks(np.arange(-8, 2, 2))\n",
    "ax6.text(0.005, 0.82, f\"R$={r2_mat:.3f}\", transform=ax6.transAxes)\n",
    "\n",
    "ax7.plot(x_gini_nas, gini_nas)\n",
    "ax7.plot(x_gini_mat, gini_mat)\n",
    "ax7.set_ylim([0.3, 1])\n",
    "ax7.set_xlabel('Position in primary rRNA transcript')\n",
    "ax7.set_ylabel('Gini index')\n",
    "dscatter_plot(gini_mat, gini_nas, ax=ax8, order=True, alpha=0.15, markersize=5, edgecolors='none', rasterized=True, cmap=cmap)\n",
    "ax8.plot([0, 1], [0,  1], '--k')\n",
    "ax8.set_aspect('equal', adjustable=\"datalim\")\n",
    "ax8.set_xlim([0.3, 1])\n",
    "ax8.set_ylim([0.3, 1])\n",
    "ax8.set_xticks(np.arange(0.2, 1.1, 0.2))\n",
    "ax8.set_yticks(np.arange(0.2, 1.1, 0.2))\n",
    "\n",
    "plt.setp([ax1, ax2, ax3], ylim=[-0.05, 1.05])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18b8176f-eda7-4944-8b5b-9adf8127095a",
   "metadata": {},
   "source": [
    "## Figure S1a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9752af59-a811-4ef0-9884-bbf1a463131d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data = pd.read_csv('../data/RQ19739-RQ19825-cD1000_Electropherogram.csv')\n",
    "\n",
    "# plot\n",
    "srt = 350\n",
    "end = 600\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(8, 3))\n",
    "\n",
    "axs[0].plot(data['A1: Ladder'][srt:end]/np.sum(data['A1: Ladder'][srt:end]), label='DNA ladder')\n",
    "axs[1].plot(data['D3: Ladder'][srt:end]/np.sum(data['A1: Ladder'][srt:end]))\n",
    "\n",
    "# mature\n",
    "axs[0].plot(data['C2: mat_d3_1'][srt:end]/np.sum(data['C2: mat_d3_1'][srt:end]), '-k')\n",
    "axs[0].plot(data['C1: mat_wt_2'][srt:end]/np.sum(data['C1: mat_wt_2'][srt:end]), '-k')\n",
    "axs[0].plot(data['E2: mat_d3_3'][srt:end]/np.sum(data['E2: mat_d3_3'][srt:end]), '-k', label='libraries')\n",
    "axs[0].set_title('DMS-MaPseq library examples')\n",
    "axs[0].legend()\n",
    "\n",
    "# nascent\n",
    "axs[1].plot(data['C5: pro_s1_3'][srt:end]/np.sum(data['C5: pro_s1_3'][srt:end]), '-k')\n",
    "axs[1].plot(data['D5: pro_d7_1'][srt:end]/np.sum(data['D5: pro_d7_1'][srt:end]), '-k')\n",
    "axs[1].plot(data['E5: pro_d7_2'][srt:end]/np.sum(data['E5: pro_d7_2'][srt:end]), '-k')\n",
    "axs[1].set_title('CoSTseq library examples')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aff5c81f-8b57-4427-ae06-3127834eb8e1",
   "metadata": {},
   "source": [
    "## Figure S1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05b75c19-a4d6-42ef-91f1-6660ffee4101",
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_pro = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_2_mRNA_agg.pkl', 'wild_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_wild_3_mRNA_agg.pkl', 'wild_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp7_1_mRNA_agg.pkl', 'dbp7_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp7_2_mRNA_agg.pkl', 'dbp7_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp7_3_mRNA_agg.pkl', 'dbp7_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp3_1_mRNA_agg.pkl', 'dbp3_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp3_2_mRNA_agg.pkl', 'dbp3_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_dbp3_3_mRNA_agg.pkl', 'dbp3_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_spt4_1_mRNA_agg.pkl', 'spt4_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_spt4_2_mRNA_agg.pkl', 'spt4_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_spt4_3_mRNA_agg.pkl', 'spt4_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_stm1_1_mRNA_agg.pkl', 'stm1_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_stm1_2_mRNA_agg.pkl', 'stm1_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_stm1_3_mRNA_agg.pkl', 'stm1_R3', genome)], 'all_pro', genome, min_cov=900)\n",
    "\n",
    "dms_mat = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_2_mRNA_agg.pkl', 'wild_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_wild_3_mRNA_agg.pkl', 'wild_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp7_1_mRNA_agg.pkl', 'dbp7_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp7_2_mRNA_agg.pkl', 'dbp7_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp7_3_mRNA_agg.pkl', 'dbp7_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp3_1_mRNA_agg.pkl', 'dbp3_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp3_2_mRNA_agg.pkl', 'dbp3_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_dbp3_3_mRNA_agg.pkl', 'dbp3_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_spt4_1_mRNA_agg.pkl', 'spt4_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_spt4_2_mRNA_agg.pkl', 'spt4_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_spt4_3_mRNA_agg.pkl', 'spt4_R3', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_stm1_1_mRNA_agg.pkl', 'stm1_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_stm1_2_mRNA_agg.pkl', 'stm1_R2', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_stm1_3_mRNA_agg.pkl', 'stm1_R3', genome)], 'all_mat', genome, min_cov=900)\n",
    "\n",
    "unt_mat = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_unwt_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/mat_dms/mat_unwt_2_mRNA_agg.pkl', 'wild_R2', genome)], 'all_mat', genome, min_cov=900)\n",
    "                           \n",
    "unt_pro = combine_profiles([ DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_unwt_1_mRNA_agg.pkl', 'wild_R1', genome), \n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_unwt_2_mRNA_agg.pkl', 'wild_R2', genome),\n",
    "                             DMS_Profile('/Users/leo/Documents/repos/PRODMSMaPseq/data/muts/pro_dms/pro_unwt_3_mRNA_agg.pkl', 'wild_R2', genome)], 'all_mat', genome, min_cov=900)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82f16a8-6bae-4ba0-96f0-df223291abf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate mutation rates for individual nucleotides\n",
    "\n",
    "def apply_mask(profile, mask):\n",
    "    arr = np.hstack([p[msk] for p, msk in zip(profile, mask)])\n",
    "    return(arr)\n",
    "\n",
    "def comb_mask(mask1, mask2):\n",
    "    return([i&j for i, j in zip(mask1, mask2)])\n",
    "\n",
    "# individual nt masks\n",
    "U = genome.mask_U()\n",
    "A = genome.mask_A()\n",
    "G = genome.mask_G()\n",
    "C = genome.mask_C()\n",
    "    \n",
    "# treated\n",
    "pro_U = np.hstack([apply_mask(dms_pro.rat_p, comb_mask(U, dms_pro.mask_p)), apply_mask(dms_pro.rat_m, comb_mask(A, dms_pro.mask_m))])\n",
    "pro_A = np.hstack([apply_mask(dms_pro.rat_p, comb_mask(A, dms_pro.mask_p)), apply_mask(dms_pro.rat_m, comb_mask(U, dms_pro.mask_m))])\n",
    "pro_G = np.hstack([apply_mask(dms_pro.rat_p, comb_mask(G, dms_pro.mask_p)), apply_mask(dms_pro.rat_m, comb_mask(C, dms_pro.mask_m))])\n",
    "pro_C = np.hstack([apply_mask(dms_pro.rat_p, comb_mask(C, dms_pro.mask_p)), apply_mask(dms_pro.rat_m, comb_mask(G, dms_pro.mask_m))])\n",
    "\n",
    "mat_U = np.hstack([apply_mask(dms_mat.rat_p, comb_mask(U, dms_mat.mask_p)), apply_mask(dms_mat.rat_m, comb_mask(A, dms_mat.mask_m))])\n",
    "mat_A = np.hstack([apply_mask(dms_mat.rat_p, comb_mask(A, dms_mat.mask_p)), apply_mask(dms_mat.rat_m, comb_mask(U, dms_mat.mask_m))])\n",
    "mat_G = np.hstack([apply_mask(dms_mat.rat_p, comb_mask(G, dms_mat.mask_p)), apply_mask(dms_mat.rat_m, comb_mask(C, dms_mat.mask_m))])\n",
    "mat_C = np.hstack([apply_mask(dms_mat.rat_p, comb_mask(C, dms_mat.mask_p)), apply_mask(dms_mat.rat_m, comb_mask(G, dms_mat.mask_m))])\n",
    "\n",
    "# untreated\n",
    "unp_U = np.hstack([apply_mask(unt_pro.rat_p, comb_mask(U, unt_pro.mask_p)), apply_mask(unt_pro.rat_m, comb_mask(A, unt_pro.mask_m))])\n",
    "unp_A = np.hstack([apply_mask(unt_pro.rat_p, comb_mask(A, unt_pro.mask_p)), apply_mask(unt_pro.rat_m, comb_mask(U, unt_pro.mask_m))])\n",
    "unp_G = np.hstack([apply_mask(unt_pro.rat_p, comb_mask(G, unt_pro.mask_p)), apply_mask(unt_pro.rat_m, comb_mask(C, unt_pro.mask_m))])\n",
    "unp_C = np.hstack([apply_mask(unt_pro.rat_p, comb_mask(C, unt_pro.mask_p)), apply_mask(unt_pro.rat_m, comb_mask(G, unt_pro.mask_m))])\n",
    "\n",
    "unm_U = np.hstack([apply_mask(unt_mat.rat_p, comb_mask(U, unt_mat.mask_p)), apply_mask(unt_mat.rat_m, comb_mask(A, unt_mat.mask_m))])\n",
    "unm_A = np.hstack([apply_mask(unt_mat.rat_p, comb_mask(A, unt_mat.mask_p)), apply_mask(unt_mat.rat_m, comb_mask(U, unt_mat.mask_m))])\n",
    "unm_G = np.hstack([apply_mask(unt_mat.rat_p, comb_mask(G, unt_mat.mask_p)), apply_mask(unt_mat.rat_m, comb_mask(C, unt_mat.mask_m))])\n",
    "unm_C = np.hstack([apply_mask(unt_mat.rat_p, comb_mask(C, unt_mat.mask_p)), apply_mask(unt_mat.rat_m, comb_mask(G, unt_mat.mask_m))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ae5ae0-60cc-4061-aa11-ba39734fe784",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot mutation rates\n",
    "\n",
    "fig, axs = plt.subplots(figsize=(8, 4))\n",
    "sample_list = [pro_G, pro_U, pro_A, pro_C, mat_G, mat_U, mat_A, mat_C, unp_G, unp_U, unp_A, unp_C, unm_G, unm_U, unm_A, unm_C]\n",
    "axs.violinplot([np.log10(i) for i in sample_list], showextrema=False)\n",
    "c = 'gray'\n",
    "axs.boxplot([np.log10(i) for i in sample_list], showfliers=False, widths=0.17, patch_artist=True, boxprops=dict(facecolor=c, color=c), capprops=dict(color=c), whiskerprops=dict(color=c))\n",
    "\n",
    "plt.xticks(np.arange(1, 17, 1), ['G', 'U', 'A', 'C', 'G', 'U', 'A', 'C','G', 'U', 'A', 'C','G', 'U', 'A', 'C'])\n",
    "plt.ylabel('log10(mutation rate)')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4349ea49-5eaf-4f2f-929f-9812d1b3062b",
   "metadata": {},
   "source": [
    "## Figure S1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "792c1f60-ce67-42e4-b6ad-3f5643b327c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.dms_utils import plot_structure, get_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbaca491-762b-4418-9f54-8bb1c7cbddaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot DMS reactivities for high correlation regions\n",
    "\n",
    "# 18S 21ES6cb\n",
    "fig, axs = plt.subplots(2, 1, figsize=[8, 2], sharex=True)\n",
    "coords = (1339, 1443)\n",
    "x_18, y_nas_18, vec_norm_nas_18 = wild_nas.plot_profile(coords[0], coords[1], genome_rrna, axs[0], cmap_loc='../cmap.txt')\n",
    "axs[0].set_ylim([0, 4])\n",
    "axs[0].set_title('18S 21ES6cb')\n",
    "axs[0].set_ylabel('DMS reactivity')\n",
    "x_18, y_rib_18, vec_norm_rib_18 = wild_rib.plot_profile(coords[0], coords[1], genome_rrna, axs[1], cmap_loc='../cmap.txt')\n",
    "axs[1].set_ylim([0, 10])\n",
    "axs[1].set_xlabel('Position in primary rRNA transcript')\n",
    "plt.show()\n",
    "\n",
    "# 25S 63ES27a\n",
    "fig, axs = plt.subplots(2, 1, figsize=[8, 2], sharex=True)\n",
    "coords = (5225, 5301)\n",
    "coords_tru_18 = get_coords(genome_rrna.seq[coords[0]:coords[1]], dbr_full[coords[0]:coords[1]])\n",
    "x_25, y_nas_25, vec_norm_nas_25 = wild_nas.plot_profile(coords[0], coords[1], genome_rrna, axs[0], cmap_loc='../cmap.txt')\n",
    "axs[0].set_ylim([0, 4])\n",
    "axs[0].set_title('25S 63ES27a')\n",
    "axs[0].set_ylabel('DMS reactivity')\n",
    "x_25, y_rib_25, vec_norm_rib_25 = wild_rib.plot_profile(coords[0], coords[1], genome_rrna, axs[1], cmap_loc='../cmap.txt')\n",
    "axs[1].set_ylim([0, 10])\n",
    "axs[1].set_xlabel('Position in primary rRNA transcript')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685b0cbb-7def-4411-aa1c-7a73db492079",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot RNA structures for high correlation regions\n",
    "\n",
    "fig, axs = plt.subplots(2, 2, figsize=(6, 10))\n",
    "fig.patch.set_facecolor('white')\n",
    "axs = axs.flatten()\n",
    "\n",
    "# 18S 21ES6cb\n",
    "coords_tru_18 = get_coords(genome_rrna.seq[1339:1443], dbr_full[1339:1443])\n",
    "plot_structure(coords_tru_18, genome_rrna.seq[1339:1443], y_rib_18*vec_norm_rib_18, (~genome_rrna.mask_G()&wild_nas.mask)[1339:1443], axs=axs[0], circle_size=40, text_size=6, cmap_loc='../cmap.txt')\n",
    "plot_structure(coords_tru_18, genome_rrna.seq[1339:1443], y_nas_18*vec_norm_nas_18, (~genome_rrna.mask_G()&wild_nas.mask)[1339:1443], axs=axs[1], circle_size=40, text_size=6, cmap_loc='../cmap.txt')\n",
    "axs[0].set_title('mature')\n",
    "axs[1].set_title('nascent')\n",
    "\n",
    "# 25S 63ES27a\n",
    "coords_tru_25 = get_coords(genome_rrna.seq[5225:5301], dbr_full[5225:5301])\n",
    "plot_structure(coords_tru_25, genome_rrna.seq[5225:5301], y_rib_25*vec_norm_rib_25, (~genome_rrna.mask_G()&wild_rib.mask)[5225:5301], axs=axs[2], circle_size=40, text_size=6, cmap_loc='../cmap.txt')\n",
    "plot_structure(coords_tru_25, genome_rrna.seq[5225:5301], y_nas_25*vec_norm_nas_25, (~genome_rrna.mask_G()&wild_rib.mask)[5225:5301], axs=axs[3], circle_size=40, text_size=6, cmap_loc='../cmap.txt')\n",
    "axs[2].set_title('mature')\n",
    "axs[3].set_title('nascent')\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
